exchange-v2/contracts/ExchangeV2Core.sol

```diff
@@ -51,54 +56,67 @@ abstract contract ExchangeV2Core is Initializable, OwnableUpgradeable, AssetMatc
+        (LibDeal.Data memory leftDealData, bool fillLeft) = LibOrderData.parse(orderLeft);
+        (LibDeal.Data memory rightDealData, bool fillRight) = LibOrderData.parse(orderRight);
 
-        LibFill.FillResult memory newFill = getFillSetNew(orderLeft, orderRight, leftOrderKeyHash, rightOrderKeyHash, leftOrderData, rightOrderData);
+        LibFill.FillResult memory newFill = getFillSetNew(orderLeft, orderRight, LibOrder.hashKey(orderLeft), LibOrder.hashKey(orderRight), fillLeft, fillRight);
 
-        (uint totalMakeValue, uint totalTakeValue) = doTransfers(makeMatch, takeMatch, newFill, orderLeft, orderRight, leftOrderData, rightOrderData);
+        runTransfers(makeMatch, takeMatch, newFill, leftDealData, rightDealData, orderLeft.maker, orderRight.maker);
+        emit Match(LibOrder.hashKey(orderLeft), LibOrder.hashKey(orderRight), orderLeft.maker, orderRight.maker, newFill.rightValue, newFill.leftValue, makeMatch, takeMatch);
+    }
+
+    function runTransfers(
+        LibAsset.AssetType memory makeMatch,
+        LibAsset.AssetType memory takeMatch,
+        LibFill.FillResult memory newFill,
+        LibDeal.Data memory leftDealData,
+        LibDeal.Data memory rightDealData,
+        address leftMaker,
+        address rightMaker
+    ) internal {
+        /*if on of assetClass == ETH, need to transfer ETH to RaribleTransferManager contract before run method doTransfers*/
         if (makeMatch.assetClass == LibAsset.ETH_ASSET_CLASS) {
-            require(takeMatch.assetClass != LibAsset.ETH_ASSET_CLASS);
-            require(msg.value >= totalMakeValue, "not enough eth");
-            if (msg.value > totalMakeValue) {
-                address(msg.sender).transferEth(msg.value.sub(totalMakeValue));
-            }
+            require(takeMatch.assetClass != LibAsset.ETH_ASSET_CLASS, "try transfer eth<->eth");
+            require(msg.value > 0, "eth == 0");
+            address(raribleTransferManager).transferEth(msg.value);
         } else if (takeMatch.assetClass == LibAsset.ETH_ASSET_CLASS) {
-            require(msg.value >= totalTakeValue, "not enough eth");
-            if (msg.value > totalTakeValue) {
-                address(msg.sender).transferEth(msg.value.sub(totalTakeValue));
-            }
+            require(msg.value > 0, "eth == 0");
+            address(raribleTransferManager).transferEth(msg.value);
         }
-        emit Match(leftOrderKeyHash, rightOrderKeyHash, orderLeft.maker, orderRight.maker, newFill.rightValue, newFill.leftValue, makeMatch, takeMatch);
+        LibAsset.Asset memory makeMatchAsset;
+        LibAsset.Asset memory takeMatchAsset;
+        makeMatchAsset.assetType = makeMatch;
+        makeMatchAsset.value = newFill.leftValue;
+        takeMatchAsset.assetType = takeMatch;
+        takeMatchAsset.value = newFill.rightValue;
+        ITransferManager(raribleTransferManager).doTransfers(makeMatchAsset, takeMatchAsset, leftDealData, rightDealData, leftMaker, rightMaker, _msgSender());
```

Comment:
we could probably do something like doTransfers(LibAsset.Asset(makeMatch, newFill.leftValue))... if it saves gas
i mean send it without creating extra variables

Reply:
agree, done